cockatrice/src/interface/widgets/visual_database_display/visual_database_display_format_legality_filter_widget.cpp

```diff
@@ -0,0 +1,205 @@
+{
+    int maxCount = 1;
+    for (auto it = allFormatsWithCount.begin(); it != allFormatsWithCount.end(); ++it) {
+        maxCount = qMax(maxCount, it.value());
+    }
+    return maxCount;
+}
+
+void VisualDatabaseDisplayFormatLegalityFilterWidget::handleFormatToggled(const QString &format, bool active)
+{
+    activeFormats[format] = active;
+
+    if (formatButtons.contains(format)) {
+        formatButtons[format]->setChecked(active);
+    }
+
+    updateFormatFilter();
+}
+
+void VisualDatabaseDisplayFormatLegalityFilterWidget::updateFormatFilter()
+{
+    // Clear existing filters related to main type
+    filterModel->blockSignals(true);
+    filterModel->filterTree()->blockSignals(true);
+    filterModel->clearFiltersOfType(CardFilter::Attr::AttrFormat);
+
+    if (exactMatchMode) {
+        // Exact Match: Only selected main types are allowed
+        QSet<QString> selectedTypes;
+        for (const auto &type : activeFormats.keys()) {
+            if (activeFormats[type]) {
+                selectedTypes.insert(type);
+            }
+        }
+
+        if (!selectedTypes.isEmpty()) {
+            // Require all selected types (TypeAnd)
+            for (const auto &type : selectedTypes) {
+                QString typeString = type;
+                filterModel->addFilter(
+                    new CardFilter(typeString, CardFilter::Type::TypeAnd, CardFilter::Attr::AttrFormat));
+            }
+
+            // Exclude any other types (TypeAndNot)
+            for (const auto &type : formatButtons.keys()) {
+                if (!selectedTypes.contains(type)) {
+                    QString typeString = type;
+                    filterModel->addFilter(
+                        new CardFilter(typeString, CardFilter::Type::TypeAndNot, CardFilter::Attr::AttrFormat));
+                }
+            }
+        }
+    } else {
+        // Default Includes Mode (TypeOr) - match any selected main types
+        for (const auto &type : activeFormats.keys()) {
+            if (activeFormats[type]) {
+                QString typeString = type;
+                filterModel->addFilter(
+                    new CardFilter(typeString, CardFilter::Type::TypeAnd, CardFilter::Attr::AttrFormat));
+            }
+        }
+    }
+
+    filterModel->blockSignals(false);
+    filterModel->filterTree()->blockSignals(false);
+
+    emit filterModel->filterTree()->changed();
+    emit filterModel->layoutChanged();
+}
+
+void VisualDatabaseDisplayFormatLegalityFilterWidget::updateFilterMode(bool checked)
+{
+    exactMatchMode = checked;
+    toggleButton->setText(exactMatchMode ? tr("Mode: Exact Match") : tr("Mode: Includes"));
+    updateFormatFilter();
+}
+
+void VisualDatabaseDisplayFormatLegalityFilterWidget::syncWithFilterModel()
+{
+    // Temporarily block signals for each button to prevent toggling while updating button states
+    for (auto it = formatButtons.begin(); it != formatButtons.end(); ++it) {
```

Comment:
is there a specific reason to use an iterator instead of simply using an expression on formatButtons.values() here?